ctsit/nacculator
# Add check_characters function to redcap2nacc

**What does this change do?** _Please be clear and concise._

This feature adds a "check_characters" function (beginning line 58) to redcap2nacc that searches the "Char" fields in each form. If it finds characters incompatible with NACC ( ' " & % ) it returns an error and skips to the next ptid. In order to differentiate "Char" fields from "Num" fields, I added a variable to the __init__.py file (in the uds3 directory) called "typename" on line 53. 
I also deleted two unused variables "pattern" and "e" as per the linter's request and made sure the program still ran without them.


**Why was this change made?** _Including an issue number is sufficient. Otherwise, briefly explain the benefit of the change._

[Issue 86](https://github.com/ctsit/nacculator/issues/86)


## Verification

_List the steps needed to make sure this thing works. Be precise._

1. From the command-line, run ... (ran using vscode debugger with data.csv, which was made from my collection of REDCap sample data)
sample:
ptid,redcap_event_name,redcap_data_access_group,formver,adcid,visitmo,visitday,visityr,visitnum,initials,
1,initial_visit_year_arm_1,gainesville_clinic,3,1,9,26,2019,1,se,

2. The output should look like this:  A regular data.txt redcap2nacc output file, e.g. "IL B1L 3   1  1          9  26 2019 1   se" as the header for the first form

3. If you pass _some bad value_, the output should look like this: 
[SKIP] Error for ptid : 10000
'FIELDNAME' is 'VA&L"UE', which has invalid character(s) " & . This field can have any text or numbers, but cannot include single quotes ', double quotes ", ampersands & or percentage signs % 

4. Pylinter returns no errors or warnings, and I ran sample data through redcap2nacc with ivp, fvp, np, lbd-ivp, and lbd-fvp formats. The output of each file did not change from last time, when I finished the LBD module, so it is behaving as expected. The check_characters function is capable of finding any combination of character errors in each field, and the number of times each offending character appears in the string. 
I have also run unit tests that confirm: the function does find and return accurate character errors; the function is able to find multiple different characters, and also multiple instances of the same character; the function returns an organized list of found characters that can be joined into one output message.
The unit test file is under the tests directory as "invalid_character_checker_test.py"

- Verify the thing does this: Successfully finds any of the four invalid characters in a string.
- Verify the thing does not do that: Does not return false positives or negatives due to backslashes, nor errors due to multiple invalid characters in one string.



## Affirmations

All of these should have a check by them. Any exception requires an explanation.

* [+] I matched the style of the existing code.
* [+] I added and updated any relevant documentation (inline, `README`, `CHANGELOG`, and such).
* [+] I used Python's type hinting.
* [+] I ran the automated tests and ensured they **ALL** passed.
* [+] I ran the linter and ensured my changes have not introduced **ANY** warnings or errors.
* [+] I have made an effort to minimize code changes and have not included any cruft (preference files, *.pyc files, old comments, print-debugging, unused variables).
* [+] I have made an effort maintain a clear commit history (haven't merged other branches or rebased improperly).
* [+] I have written the code myself or have given credit where credit is due.
